src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs

```diff
@@ -203,19 +203,31 @@ public virtual string GetVirtualPath(VirtualPathContext context)
             // Validate that the target can accept these values.
             var childContext = CreateChildVirtualPathContext(context, values.AcceptedValues);
 
-            var path = _target.GetVirtualPath(childContext);
-            if (path != null)
+            var pathData = _target.GetVirtualPath(childContext);
+            if (pathData != null)
             {
                 // If the target generates a value then that can short circuit.
-                return path;
+                return pathData;
             }
 
             // If we can produce a value go ahead and do it, the caller can check context.IsBound
             // to see if the values were validated.
-            path = _binder.BindValues(values.AcceptedValues);
+            var tempDataTokens = new RouteValueDictionary();
+            if (DataTokens != null)
+            {
+                foreach (var dataToken in DataTokens)
+                {
+                    tempDataTokens.Add(dataToken.Key, dataToken.Value);
+                }
+            }
+            pathData = new VirtualPathData(
+                router: this,
+                virtualPath: _binder.BindValues(values.AcceptedValues),
+                dataTokens: tempDataTokens);
```

Comment:
After I made a change as you suggested. I looked into GetVirtualPath method again to add an unit test on it, but ..it seems like _binder.BindValues won't return null in this case because that chance was already filtered out in the following codes:
https://github.com/aspnet/Routing/blob/dev/src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs#L185-L190
Let me know if I missed anything here.